SimG4Core/Geometry/src/DD4hep_DDG4Builder.cc

```diff
@@ -51,27 +55,37 @@ G4VPhysicalVolume *DDG4Builder::BuildGeometry(SensitiveDetectorCatalog &catalog)
     }
   }
 
+  // ADD ALL SELECTED G4 LOGICAL VOLUMES TO SENSITIVE DETECTORS CATALOGUE
   for (auto const &it : dd4hepVec) {
+    // Sensitive detector info
+    const G4String &sensitiveDetectorG4Name = it.first->GetName();
     auto sClassName = it.second->strValue("SensitiveDetector");
     auto sROUName = it.second->strValue("ReadOutName");
-    auto fff = it.first->GetName();
-    catalog.insert({sClassName.data(), sClassName.size()}, {sROUName.data(), sROUName.size()}, fff);
-
-    auto reflectedG4LogicalVolumeName = fff + "_refl";
-    bool hasReflectedVolumeInStore = false;
-    G4LogicalVolumeStore *theStore = G4LogicalVolumeStore::GetInstance();
-    for (const auto &lv : *theStore) {
-      if (lv->GetName().find(reflectedG4LogicalVolumeName) != std::string::npos) {
-        hasReflectedVolumeInStore = true;
-      }
-    }
-    if (hasReflectedVolumeInStore) {
+    // Add to catalogue
+    catalog.insert({sClassName.data(), sClassName.size()}, {sROUName.data(), sROUName.size()}, sensitiveDetectorG4Name);
+
+    edm::LogVerbatim("SimG4CoreApplication") << " DDG4SensitiveConverter: Sensitive " << sensitiveDetectorG4Name
+                                             << " Class Name " << sClassName << " ROU Name " << sROUName;
+
+    // Reflected sensors also need to be added to the senstive detectors catalogue!
+    // Similar treatment here with DD4hep, as what was done for old DD.
+    const G4String &sensitiveDetectorG4ReflectedName = sensitiveDetectorG4Name + "_refl";
+
+    const G4LogicalVolumeStore *const allG4LogicalVolumes = G4LogicalVolumeStore::GetInstance();
+    const bool hasG4ReflectedVolume =
+        std::find_if(
+            allG4LogicalVolumes->begin(), allG4LogicalVolumes->end(), [&](G4LogicalVolume *const aG4LogicalVolume) {
+              return (aG4LogicalVolume->GetName() == sensitiveDetectorG4ReflectedName);
+            }) != allG4LogicalVolumes->end();
+    if (hasG4ReflectedVolume) {
+      // Add reflected sensitive detector to catalogue
       catalog.insert(
-          {sClassName.data(), sClassName.size()}, {sROUName.data(), sROUName.size()}, reflectedG4LogicalVolumeName);
-    }
+          {sClassName.data(), sClassName.size()}, {sROUName.data(), sROUName.size()}, sensitiveDetectorG4ReflectedName);
 
-    edm::LogVerbatim("SimG4CoreApplication")
-        << " DDG4SensitiveConverter: Sensitive " << fff << " Class Name " << sClassName << " ROU Name " << sROUName;
+      edm::LogVerbatim("SimG4CoreApplication")
```

Comment:
@ghugo83 , it may be too much to add the same printout as in line 67 ?

Reply:
@civanch This is the printout of the reflected sensors when they exist, while the printout at line 67 is the printout of the non-reflected sensors.